File: scripts/generate-test-chunks.js
#!/usr/bin/env node
/**
 * Generate test video chunks for upload infrastructure testing.
 *
 * Creates mock MP4 files with the correct naming convention and size.
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const os = require('os');

/**
 * Create a test chunk file
 */
function createTestChunk(outputDir, recordingId, chunkIndex, sizeMB = 50) {
    // Create output directory
    const chunkDir = path.join(outputDir, recordingId);
    fs.mkdirSync(chunkDir, { recursive: true });

    // Generate filename (1-based for display, 0-based internally)
    const filename = `part-${String(chunkIndex + 1).padStart(4, '0')}.mp4`;
    const filepath = path.join(chunkDir, filename);

    console.log(`Creating ${filename} (${sizeMB}MB)...`);

    // Create file with random data
    const sizeBytes = sizeMB * 1024 * 1024;
    const buffer = crypto.randomBytes(sizeBytes);
    fs.writeFileSync(filepath, buffer);

    // Calculate SHA-256 checksum
    console.log(`Calculating checksum for ${filename}...`);
    const hash = crypto.createHash('sha256');
    hash.update(buffer);
    const checksum = hash.digest('hex');

    // Generate chunk metadata
    const chunkId = `${recordingId}-chunk-${String(chunkIndex).padStart(4, '0')}`;

    const metadata = {
        chunkId: chunkId,
        filePath: filepath,
        sizeBytes: sizeBytes,
        checksum: checksum,
        durationSeconds: 60.0,
        index: chunkIndex,
        recordingId: recordingId
    };

    console.log(`✅ Created: ${filepath}`);
    console.log(`   Chunk ID: ${chunkId}`);
    console.log(`   Size: ${sizeMB}MB (${sizeBytes.toLocaleString()} bytes)`);
    console.log(`   Checksum: ${checksum.substring(0, 16)}...`);
    console.log();

    return metadata;
}

function main() {
    // Configuration
    const outputDir = path.join(os.homedir(), 'Library', 'Caches', 'MeetingRecorder');
    const recordingId = 'test-rec-001';

    // Parse command line arguments
    const numChunks = parseInt(process.argv[2]) || 3;
    const sizeMB = parseInt(process.argv[3]) || 50;

    console.log(`Generating ${numChunks} test chunks of ${sizeMB}MB each...`);
    console.log(`Output directory: ${outputDir}/${recordingId}`);
    console.log(`Recording ID: ${recordingId}`);
    console.log();

    // Create chunks
    const chunks = [];
    for (let i = 0; i < numChunks; i++) {
        const chunk = createTestChunk(outputDir, recordingId, i, sizeMB);
        chunks.push(chunk);
    }

    // Summary
    console.log('='.repeat(60));
    console.log(`✅ Generated ${chunks.length} test chunks`);
    console.log(`📁 Location: ${outputDir}/${recordingId}`);
    console.log(`💾 Total size: ${chunks.length * sizeMB}MB`);
    console.log();
    console.log('Chunk files:');
    chunks.forEach(chunk => {
        console.log(`  - ${path.basename(chunk.filePath)}`);
    });
    console.log();
    console.log('To upload these chunks, use the S3Uploader in your Swift tests.');
    console.log();
}

// Run
console.log('🎬 Test Chunk Generator');
console.log();

try {
    main();
} catch (error) {
    console.error(`\n❌ Error: ${error.message}`);
    process.exit(1);
}

File: scripts/generate-test-chunks.swift
#!/usr/bin/env swift

import Foundation
import CryptoKit

/**
 Generate test video chunks for upload infrastructure testing.

 Creates mock MP4 files with the correct naming convention and size.
 */

func createTestChunk(
    outputDir: URL,
    recordingId: String,
    chunkIndex: Int,
    sizeMB: Int = 50
) throws -> [String: Any] {
    // Create output directory
    let chunkDir = outputDir.appendingPathComponent(recordingId, isDirectory: true)
    try FileManager.default.createDirectory(at: chunkDir, withIntermediateDirectories: true)

    // Generate filename (1-based for display, 0-based internally)
    let filename = String(format: "part-%04d.mp4", chunkIndex + 1)
    let filepath = chunkDir.appendingPathComponent(filename)

    print("Creating \(filename) (\(sizeMB)MB)...")

    // Create file with random data
    let sizeBytes = sizeMB * 1024 * 1024
    var data = Data(count: sizeBytes)
    let result = data.withUnsafeMutableBytes { bytes in
        SecRandomCopyBytes(kSecRandomDefault, sizeBytes, bytes.baseAddress!)
    }

    guard result == errSecSuccess else {
        throw NSError(domain: "TestChunkGenerator", code: 1, userInfo: [
            NSLocalizedDescriptionKey: "Failed to generate random data"
        ])
    }

    try data.write(to: filepath, options: .atomic)

    // Calculate SHA-256 checksum
    print("Calculating checksum for \(filename)...")
    let hash = SHA256.hash(data: data)
    let checksum = hash.compactMap { String(format: "%02x", $0) }.joined()

    // Generate chunk metadata
    let chunkId = "\(recordingId)-chunk-\(String(format: "%04d", chunkIndex))"

    let metadata: [String: Any] = [
        "chunkId": chunkId,
        "filePath": filepath.path,
        "sizeBytes": sizeBytes,
        "checksum": checksum,
        "durationSeconds": 60.0,
        "index": chunkIndex,
        "recordingId": recordingId
    ]

    print("✅ Created: \(filepath.path)")
    print("   Chunk ID: \(chunkId)")
    print("   Size: \(sizeMB)MB (\(sizeBytes.formatted()) bytes)")
    print("   Checksum: \(checksum.prefix(16))...")
    print()

    return metadata
}

func main() throws {
    // Configuration
    let homeDir = FileManager.default.homeDirectoryForCurrentUser
    let outputDir = homeDir
        .appendingPathComponent("Library", isDirectory: true)
        .appendingPathComponent("Caches", isDirectory: true)
        .appendingPathComponent("MeetingRecorder", isDirectory: true)
    let recordingId = "test-rec-001"

    // Parse command line arguments
    let args = CommandLine.arguments
    let numChunks = args.count > 1 ? Int(args[1]) ?? 3 : 3
    let sizeMB = args.count > 2 ? Int(args[2]) ?? 50 : 50

    print("Generating \(numChunks) test chunks of \(sizeMB)MB each...")
    print("Output directory: \(outputDir.path)/\(recordingId)")
    print("Recording ID: \(recordingId)")
    print()

    // Create chunks
    var chunks: [[String: Any]] = []
    for i in 0..<numChunks {
        let chunk = try createTestChunk(
            outputDir: outputDir,
            recordingId: recordingId,
            chunkIndex: i,
            sizeMB: sizeMB
        )
        chunks.append(chunk)
    }

    // Summary
    print(String(repeating: "=", count: 60))
    print("✅ Generated \(chunks.count) test chunks")
    print("📁 Location: \(outputDir.path)/\(recordingId)")
    print("💾 Total size: \(chunks.count * sizeMB)MB")
    print()
    print("Chunk files:")
    for chunk in chunks {
        if let path = chunk["filePath"] as? String {
            print("  - \(URL(fileURLWithPath: path).lastPathComponent)")
        }
    }
    print()
    print("To upload these chunks, use the S3Uploader in your Swift tests.")
    print()
}

// Run
print("🎬 Test Chunk Generator")
print()

do {
    try main()
} catch {
    print("\n❌ Error: \(error.localizedDescription)")
    exit(1)
}
